pycyphal/dsdl/_import_hook.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (c) 2019 OpenCyphal
+# This software is distributed under the terms of the MIT License.
+
+import logging
+import sys
+import os
+from typing import Iterable, Optional, Union
+import pathlib
+from . import compile
+
+from importlib.abc import MetaPathFinder
+from importlib.util import spec_from_file_location
+
+_AnyPath = Union[str, pathlib.Path]
+
+_logger = logging.getLogger(__name__)
+
+
+class DsdlMetaFinder(MetaPathFinder):
+    def __init__(
+        self,
+        lookup_directories: Iterable[_AnyPath],
+        output_directory: _AnyPath,
+        allow_unregulated_fixed_port_id: bool,
+    ) -> None:
+        super().__init__()
+
+        _logger.debug("lookup dirs: %s", lookup_directories)
+        _logger.debug("output dir: %s", output_directory)
+
+        self.lookup_directories = list(map(str, lookup_directories))
+        self.output_directory = output_directory
+        self.allow_unregulated_fixed_port_id = allow_unregulated_fixed_port_id
+        self.root_namespace_directories: Iterable[pathlib.Path] = list()
+
+        # Build a list of root namespace directories from lookup directories.
+        # Any dir inside any of the lookup directories is considered a root namespace.
+        for dir in self.lookup_directories:
+            for namespace in pathlib.Path(dir).iterdir():
+                if namespace.is_dir():
```

Comment:
I think we should also check whether it matches the namespace name regex to avoid the inclusion of `.git` et al.

Reply:
I assume it would be best to use `check_name` from `pydsdl`, but it is private. Is there a regex for a valid namespace somewhere?

Reply:
Yes, section 3.2.2:

<img width="318" alt="image" src="https://user-images.githubusercontent.com/3298404/185419441-5fac16cf-7333-41fa-bd0c-b6fed3192821.png">

`[a-zA-Z_][a-zA-Z0-9_]*`